kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,988 @@
+    ) -> Result<StoredInFlash, ProcessBinaryError> {
+        // We only need tbf header information to get the size of app which is
+        // already loaded.
+        let test_header_slice = possible_app
+            .get(0..8)
+            .ok_or(ProcessBinaryError::NotEnoughFlash)?;
+
+        // Pass the first eight bytes to tbfheader to parse out the length of
+        // the tbf header and app. We then use those values to see if we have
+        // enough flash remaining to parse the remainder of the header.
+        let (version, header_length, _entry_length) =
+            match tock_tbf::parse::parse_tbf_header_lengths(
+                test_header_slice
+                    .try_into()
+                    .or(Err(ProcessBinaryError::TbfHeaderNotFound))?,
+            ) {
+                Ok((v, hl, el)) => (v, hl, el),
+                Err(tock_tbf::types::InitialTbfParseError::InvalidHeader(_entry_length)) => {
+                    // If we could not parse the header, then we want to skip
+                    // over this app and look for the next one.
+                    return Err(ProcessBinaryError::TbfHeaderNotFound);
+                }
+                Err(tock_tbf::types::InitialTbfParseError::UnableToParse) => {
+                    // Since Tock apps use a linked list, it is very possible
+                    // the header we started to parse is intentionally invalid
+                    // to signal the end of apps.
+                    return Ok(StoredInFlash::Empty);
+                }
+            };
+
+        // If a padding app exists at the start address satisfying MPU rules, we
+        // load the new app from here!
+        let header_flash = possible_app
+            .get(0..(header_length as usize))
+            .ok_or(ProcessBinaryError::NotEnoughFlash)?;
+
+        let tbf_header = tock_tbf::parse::parse_tbf_header(header_flash, version)?;
+
+        // If this isn't an app (i.e. it is padding).
+        if tbf_header.is_app() {
+            Ok(StoredInFlash::ValidApp(tbf_header.length() as usize))
+        } else {
+            Ok(StoredInFlash::PaddingApp)
+        }
+    }
+
+    /// Check if our new app overlaps with existing apps
+    fn check_overlap_region(
+        &self,
+        new_start_address: usize,
+        app_length: usize,
+    ) -> Result<(), (usize, ProcessLoadError)> {
+        // Find the next open process slot.
+        let new_process_count = self.find_open_process_slot().unwrap_or_default();
+        let new_process_start_address = new_start_address;
+        let new_process_end_address = new_process_start_address + app_length - 1;
+
+        self.procs.map(|procs| {
+            for (proc_index, value) in procs.iter().enumerate() {
+                if proc_index < new_process_count {
+                    let process_start_address = value.unwrap().get_addresses().flash_start;
+                    let process_end_address = value.unwrap().get_addresses().flash_end;
```

Comment:
Unfortunately this isn't going to work. The main issue is that there is no requirement that loaded processes (ie in the `PROCESSES` array) represent all stored process binaries.

What we probably need to do instead is use the `ProcessBinaries` array.